### ISSUE TITLE
PerlIO::Layer::->find fails to discover on-demand layers

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6237]
Migrated from [rt.perl.org#20460](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20460) (status was 'resolved')

Searchable as RT20460$

**Comment by p5pRT** [comment]
### From @audreyt

#### Created by autrijus@autrijus.org



$ perl \-M'open=locale' \-wedie\-ok
\-ok at \-e line 1\.

$ perl \-M'open=encoding\(big5\)' \-wedie\-ok
Unknown PerlIO layer 'encoding\(big5\)' at \-e line 0
\-ok at \-e line 1\.

$ perl \-M'open=fnording\(big5\)' \-wedie\-ok
Unknown PerlIO layer class 'fnording\(big5\)' at \-e line 0
BEGIN failed\-\-compilation aborted\.

Isn't the warning on the second case misleading and
redundant? :\-\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.0:

Configured by autrijus at Mon Jan 20 07:31:27 CST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=freebsd, osvers=freebsd, archname=i386-freebsd
    uname='freebsd 5.0-release freebsd 5.0-release #0: thu jan 16 22:16:53 gmt 2003 root@hollin.btc.adaptec.com:usrobjusrsrcsysgeneric i386 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -I/usr/local/include',
    optimize='-O',
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.2.1 [FreeBSD] 20021119 (release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lgdbm -ldb -lm -lc -lcrypt -lutil
    perllibs=-lm -lc -lcrypt -lutil
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fpic', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    MAINT18448


@INC for perl v5.8.0:
    /usr/local/lib/perl5/5.8.0/i386-freebsd
    /usr/local/lib/perl5/5.8.0
    /usr/local/lib/perl5/site_perl/5.8.0/i386-freebsd
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/autrijus
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=zh_TW.Big5
    LD_LIBRARY_PATH=/usr/local/lib
    LOGDIR (unset)
    PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/X11R6/bin:/home/autrijus/bin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/psh

```
</details>

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Autrijus Tang \<perl5\-porters@&#8203;perl\.org> writes&#8203;:
>\# New Ticket Created by  Autrijus Tang 
>\# Please include the string&#8203;:  \[perl \#20460\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=20460 >
>
>
>This is a bug report for perl from autrijus@&#8203;autrijus\.org\,
>generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>$ perl \-M'open=locale' \-wedie\-ok
>\-ok at \-e line 1\.
>
>$ perl \-M'open=encoding\(big5\)' \-wedie\-ok
>Unknown PerlIO layer 'encoding\(big5\)' at \-e line 0
>\-ok at \-e line 1\.
>
>$ perl \-M'open=fnording\(big5\)' \-wedie\-ok
>Unknown PerlIO layer class 'fnording\(big5\)' at \-e line 0
>BEGIN failed\-\-compilation aborted\.
>
>Isn't the warning on the second case misleading and
>redundant? :\-\)

I have tweaked it so now for 2nd case you get&#8203;:

nick@&#8203;bactrian 1072$ \./perl \-Ilib \-M'open=encoding\(big5\)' \-wedie\-ok
Unknown PerlIO layer 'encoding' at \-e line 0
\-ok at \-e line 1\.

Are you really complaining that PerlIO&#8203;::encoding is not autoloaded?
That would be easily fixed by passing a true value to PerlIO&#8203;::Layer&#8203;::\->find
"load" argument \(as shown in attached patch\)\.

Note you get the different message for 3rd case as only utf8/locale/encoding
default to 'IO'

nick@&#8203;bactrian 1074$ \./perl \-Ilib \-M'open=IO\,fnording\(big5\)' \-wedie\-ok
Unknown PerlIO layer 'fnording' at \-e line 0
\-ok at \-e line 1\.
nick@&#8203;bactrian 1075$

>
>\[Please do not change anything below this line\]
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\-\-\-
>Flags&#8203;:
>    category=core
>    severity=low
>\-\-\-
>Site configuration information for perl v5\.8\.0&#8203;:
>
>Configured by autrijus at Mon Jan 20 07&#8203;:31&#8203;:27 CST 2003\.
>
>Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
>  Platform&#8203;:
>    osname=freebsd\, osvers=freebsd\, archname=i386\-freebsd
>    uname='freebsd 5\.0\-release freebsd 5\.0\-release \#0&#8203;: thu jan 16 22&#8203;:16&#8203;:53 gmt 2003 root@&#8203;hollin\.btc\.adaptec\.com&#8203;:usrobjusrsrcsysgeneric i386 '
>    config\_args=''
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>    use64bitint=undef use64bitall=undef uselongdouble=undef
>    usemymalloc=n\, bincompat5005=undef
>  Compiler&#8203;:
>    cc='cc'\, ccflags ='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-I/usr/local/include'\,
>    optimize='\-O'\,
>    cppflags='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-I/usr/local/include'
>    ccversion=''\, gccversion='3\.2\.1 \[FreeBSD\] 20021119 \(release\)'\, gccosandvers=''
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>    alignbytes=4\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='cc'\, ldflags ='\-Wl\,\-E  \-L/usr/local/lib'
>    libpth=/usr/lib /usr/local/lib
>    libs=\-lgdbm \-ldb \-lm \-lc \-lcrypt \-lutil
>    perllibs=\-lm \-lc \-lcrypt \-lutil
>    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
>    gnulibc\_version=''
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
>    cccdlflags='\-DPIC \-fpic'\, lddlflags='\-shared  \-L/usr/local/lib'
>
>Locally applied patches&#8203;:
>    MAINT18448
>
>\-\-\-
>@&#8203;INC for perl v5\.8\.0&#8203;:
>    /usr/local/lib/perl5/5\.8\.0/i386\-freebsd
>    /usr/local/lib/perl5/5\.8\.0
>    /usr/local/lib/perl5/site\_perl/5\.8\.0/i386\-freebsd
>    /usr/local/lib/perl5/site\_perl/5\.8\.0
>    /usr/local/lib/perl5/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.8\.0&#8203;:
>    HOME=/home/autrijus
>    LANG \(unset\)
>    LANGUAGE \(unset\)
>    LC\_CTYPE=zh\_TW\.Big5
>    LD\_LIBRARY\_PATH=/usr/local/lib
>    LOGDIR \(unset\)
>    PATH=/sbin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/usr/games&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11R6/bin&#8203;:/home/autrijus/bin
>    PERL\_BADLANG \(unset\)
>    SHELL=/usr/local/bin/psh
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

<details><summary>Inline Patch</summary>

```diff
--- lib/open.pm.ship	Wed Jan 22 10:12:43 2003
+++ lib/open.pm	Wed Jan 22 10:10:21 2003
@@ -95,8 +95,8 @@
 		my $target = $layer;		# the layer name itself
 		$target =~ s/^(\w+)\(.+\)$/$1/;	# strip parameters
 
-		unless(PerlIO::Layer::->find($target)) {
-		    warnings::warnif("layer", "Unknown PerlIO layer '$layer'");
+		unless(PerlIO::Layer::->find($target,1)) {
+		    warnings::warnif("layer", "Unknown PerlIO layer '$target'");
 		}
 	    }
 	    push(@val,":$layer");
```
</details>

**Comment by p5pRT** [comment]
### From @jhi

Marking as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @audreyt

On Wed\, Jan 22\, 2003 at 10&#8203;:14&#8203;:47AM \-0000\, Nick Ing\-Simmons wrote&#8203;:
> Are you really complaining that PerlIO&#8203;::encoding is not autoloaded?
> That would be easily fixed by passing a true value to PerlIO&#8203;::Layer&#8203;::\->find
> "load" argument \(as shown in attached patch\)\.

Yes I am\.  So if that patch is in\, I'd be happy\. :\-\)

/Autrijus/

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Autrijus Tang \<autrijus@&#8203;autrijus\.org> writes&#8203;:
>On Wed\, Jan 22\, 2003 at 10&#8203;:14&#8203;:47AM \-0000\, Nick Ing\-Simmons wrote&#8203;:
>> Are you really complaining that PerlIO&#8203;::encoding is not autoloaded?
>> That would be easily fixed by passing a true value to PerlIO&#8203;::Layer&#8203;::\->find
>> "load" argument \(as shown in attached patch\)\.
>
>Yes I am\.  So if that patch is in\, I'd be happy\. :\-\)

It is in\. It had some koock on effects \- as the error messages changed
to be ones about "cannot load" rather than "unknown layer" in some 
of tests but I think that is fixed again now\.

\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'
